Allow user to define element other than map container for full screen control

ryanhamley opened the pull request:
fixes https://github.com/mapbox/mapbox-gl-js/issues/7544

## Launch Checklist

The intended use case is for displaying a map with an overlaid element that works with the map, such as radio buttons to toggle styles or a banner element with a legend. The current behavior of making the map container element full screen will not show these elements. I can't really think of any particular reason to disallow this behavior but let me know if there's a reason not to implement this.

Full screen with `{source: body}`
![screen shot 2018-11-02 at 12 40 43 pm](https://user-images.githubusercontent.com/4523080/47937034-ca97a380-de9c-11e8-8fdb-b3bc948800b8.png)

Default behavior (notice that the debug control in the upper left corner is not displayed)
![screen shot 2018-11-02 at 12 42 00 pm](https://user-images.githubusercontent.com/4523080/47937036-ca97a380-de9c-11e8-9819-55e4c7a3e7ab.png)

<!-- Thanks for the PR! Feel free to add or remove items from the checklist. -->

<!-- If your PR affects documentation relevant to the currently released version, please use `mb-pages` as the base branch. See https://github.com/mapbox/mapbox-gl-js/blob/master/docs/README.md#committing-and-publishing-documentation -->

 - [x] briefly describe the changes in this PR
 - [x] write tests for all new functionality
 - [x] document any changes to public APIs
 - [x] manually test the debug page

mourner commented on src/ui/control/fullscreen_control.js:
```diff
@@ -7,26 +7,36 @@ import window from '../../util/window';
+    constructor(options: Options) {
         this._fullscreen = false;
+        if (options && options.source) {
+            this._source = window.document.querySelector(options.source);
```
Our other APIs that need an HTML element reference, such as Map `container` and Marker `element`, accept an `HTMLElement` rather than a selector — I'd suggest keeping this consistent.

mollymerp commented on src/ui/control/fullscreen_control.js:
```diff
@@ -45,7 +59,7 @@ class FullscreenControl {
     onAdd(map: Map) {
         this._map = map;
-        this._mapContainer = this._map.getContainer();
+        if (!this._source) this._source = this._map.getContainer();
```
I think `container`/`mapContainer` is a clearer word for this member / option

ryanhamley commented:
Added a unit test and changed `source` option to `container` and the previous `_container` internal property to `_controlContainer` to disambiguate them.

mollymerp commented on test/unit/ui/control/fullscreen.test.js:
```diff
@@ -27,3 +27,18 @@ test('FullscreenControl does not appears when fullscreen is not enabled', (t) =>
+    const control = map._controls.find((ctrl) => {
+        return ctrl.hasOwnProperty('_fullscreen');
+    });
+    control._onClickFullscreen();
```
maybe set a spy on `body.requestFullscreen` and then check `t.calledOnce` after this?

ryanhamley replied:
I might be missing something but it doesn't seem like JSDOM supports the Full Screen API. I don't see any documentation or code that mentions it and `window.document.requestFullScreen` and `body.requestFullScreen` are both `undefined` (along with their prefixed alternatives).

mollymerp replied:
ahh ok nvm then!
